### PR TITLE
integrate ACP-118

### LINKED ISSUE #383: Use ACP-118 interface
**Context and scope**
Once [ACP-118](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/118-warp-signature-request) has been implemented we need to switch our `AppRequests` to use the new format as well as expose the `justification` field in the `raw-message` interface 

**Discussion and alternatives**

**Open questions**

### PR DESCRIPTION
**Pull request by feuGeneA**
## Why this should be merged
fixes #383

### CODE REVIEW

File: signature-aggregator/aggregator/aggregator.go [hunk 1]
```diff
@@ -121,19 +116,10 @@ func (s *SignatureAggregator) AggregateSignaturesAppRequest(
...
-		reqBytes, err = coreEthMsg.RequestToBytes(coreEthCodec, req)
-	} else {
-		req := msg.MessageSignatureRequest{
-			MessageID: unsignedMessage.ID(),
-		}
-		reqBytes, err = msg.RequestToBytes(codec, req)
-	}
+	reqBytes, err := msg.RequestToBytes(
+		msg.Codec,
+		msg.MessageSignatureRequest{MessageID: unsignedMessage.ID()},
```
Comment (iansuvak) [inline signature-aggregator/aggregator/aggregator.go:121]: This is still the legacy API. ACP-118 doesn't support requests by MessageID (although they will still be available). We should switch to using the protobuf encoding following @cam-schultz's reference implementation in avalanche-go at ava-labs/avalanchego#3218

File: signature-aggregator/aggregator/aggregator.go [hunk 2]
```diff
@@ -121,19 +118,9 @@ func (s *SignatureAggregator) AggregateSignaturesAppRequest(
...
-		}
-		reqBytes, err = coreEthMsg.RequestToBytes(coreEthCodec, req)
-	} else {
-		req := msg.MessageSignatureRequest{
-			MessageID: unsignedMessage.ID(),
-		}
-		reqBytes, err = msg.RequestToBytes(codec, req)
-	}
+	reqBytes, err := proto.Marshal(
+		&sdk.SignatureRequest{Message: unsignedMessage.Bytes()},
```
Comment (iansuvak) [inline signature-aggregator/aggregator/aggregator.go:122]: We should add justification bytes here as well as the API. In the API they should be hex encoded and since their meaning is VM specific we don't need to do any processing to them either just like with the message. We should only pass the non-zero fields since the ACP-118 spec but in the API layer we should do validation that at least one of `message`, `justification` fields is non-zero. 

I would also rename our API to use `Message` instead of `UnsignedMessage` to match the ACP-118 spec while we are at it

Comment (feuGeneA) [reply signature-aggregator/aggregator/aggregator.go:123]: addressed in https://github.com/ava-labs/awm-relayer/pull/408/commits/557262accb902b63364c1ce5e994f08071d45562

File: tests/utils/utils.go [hunk 3]
```diff
@@ -287,7 +287,7 @@ func CreateDefaultSignatureAggregatorConfig(
 	)
 	// Construct the config values for each subnet
 	return signatureaggregatorcfg.Config{
-		LogLevel: logging.Info.LowerString(),
+		LogLevel: logging.Debug.LowerString(),
```
Comment (feuGeneA) [inline tests/utils/utils.go:290]: Undo these changes once we get things working nicely.

File: signature-aggregator/aggregator/aggregator.go (hunk 2 above)
Comment (feuGeneA) [reply signature-aggregator/aggregator/aggregator.go:124]: This isn't done yet. I added the justification everywhere (in the API) _except for_ right here 😅

File: tests/signature_aggregator_api.go [hunk 4]
```diff
@@ -69,7 +69,7 @@ func SignatureAggregatorAPI(network interfaces.LocalNetwork) {
 	warpMessage := getWarpMessageFromLog(ctx, receipt, subnetAInfo)
 
 	reqBody := api.AggregateSignatureRequest{
-		UnsignedMessage: "0x" + hex.EncodeToString(warpMessage.Bytes()),
+		Message: "0x" + hex.EncodeToString(warpMessage.Bytes()),
```
Comment (feuGeneA) [inline tests/signature_aggregator_api.go:72]: add justification here too

**Comment by geoff-vball** [comment]
@feuGeneA is this PR still relevant?

**Comment by iansuvak** [comment]
@geoff-vball 

> @feuGeneA is this PR still relevant?

It hasn't been merged yet so yes it is. I will update my draft PR  (#422) that goes into this PR and then we can bring this one up to speed with `main` and then it should be ready for review

File: signature-aggregator/aggregator/aggregator.go (hunk 2 above)
Comment (feuGeneA) [reply signature-aggregator/aggregator/aggregator.go]: last piece addressed in https://github.com/ava-labs/awm-relayer/pull/408/commits/29cac3003f36112564c3545f2647bbe934609e54

**Comment by feuGeneA** [comment]
Ready for review, but not ready to merge until we can pin to a released version of avalanchego.

File: signature-aggregator/api/api.go [hunk 5]
```diff
@@ -26,8 +26,11 @@ const (
 
 // Defines a request interface for signature aggregation for a raw unsigned message.
 type AggregateSignatureRequest struct {
+	// Required: either Message or Justification must be provided
 	// Required. hex-encoded message, optionally prefixed with "0x".
```
Comment (geoff-vball) [inline signature-aggregator/api/api.go:30]:
```suggestion
	// Required: either Message or Justification must be provided.
	// hex-encoded message, optionally prefixed with "0x".
```

File: tests/utils/utils.go (hunk 3 above)
Comment (geoff-vball) [reply tests/utils/utils.go:289]: Are things working nicely now, or do we still need this?

File: relayer/application_relayer.go [hunk 6]
```diff
@@ -204,6 +204,7 @@ func (r *ApplicationRelayer) ProcessMessage(handler messages.MessageHandler) (co
 	if r.sourceWarpSignatureClient == nil {
 		signedMessage, err = r.signatureAggregator.CreateSignedMessage(
 			unsignedMessage,
+			make([]byte, 0),
```
Comment (minghinmatthewlam) [inline relayer/application_relayer.go:207]: wasn't too clear to me what this `justification` field is meant for, and why it's always empty

Comment (iansuvak) [reply relayer/application_relayer.go:207]: Justification field is part of [ACP-118](https://github.com/avalanche-foundation/ACPs/blob/main/ACPs/118-warp-signature-request/README.md) It's unused for now but is meant to be potentially extra data that would explain why  the data in the message field should be signed. It's unused for now though so that's why it's always empty

Comment (cam-schultz) [reply relayer/application_relayer.go:207]: Is there a reason we can't just pass in `nil` here? Allocating a slice that's never assigned to seems odd.

File: signature-aggregator/aggregator/aggregator.go [hunk 7]
```diff
@@ -176,19 +174,14 @@ func (s *SignatureAggregator) CreateSignedMessage(
...
-			MessageID: unsignedMessage.ID(),
-		}
-		reqBytes, err = coreEthMsg.RequestToBytes(coreEthCodec, req)
-	} else {
-		req := msg.MessageSignatureRequest{
-			MessageID: unsignedMessage.ID(),
-		}
-		reqBytes, err = msg.RequestToBytes(codec, req)
-	}
+	reqBytes := networkP2P.ProtocolPrefix(networkP2P.SignatureRequestHandlerID)
```
Comment (cam-schultz) [inline signature-aggregator/aggregator/aggregator.go:177]: Consider using `[p2p.PrefixMessage](https://github.com/ava-labs/avalanchego/blob/ab83fb41528de93c1790301cdd67a07dda9299f0/network/p2p/client.go#L79)` to construct `reqBytes`.

File: signature-aggregator/aggregator/aggregator.go [hunk 8]
```diff
@@ -557,8 +560,9 @@ func (s *SignatureAggregator) isValidSignatureResponse(
 		)
 		return blsSignatureBuf{}, false
 	}
-
-	return signature, true
+	blsSig := blsSignatureBuf{}
+	copy(blsSig[:], signature[:])
+	return blsSig, true
```
Comment (cam-schultz) [inline signature-aggregator/aggregator/aggregator.go:565]: We can just `return [bls.SignatureLength]byte(signature)` here

File: signature-aggregator/aggregator/aggregator.go (hunk 7 above)
Comment (iansuvak) [reply signature-aggregator/aggregator/aggregator.go:177]: Thanks for pointing this out. Addressed in the other PR https://github.com/ava-labs/awm-relayer/pull/463

File: signature-aggregator/aggregator/aggregator.go (hunk 8 above)
Comment (iansuvak) [reply signature-aggregator/aggregator/aggregator.go]: Addressed in #463

File: scripts/versions.sh [hunk 9]
```diff
@@ -17,12 +17,14 @@ function getDepVersion() {
 export GO_VERSION=${GO_VERSION:-$(getDepVersion go)}
 
 # Don't export them as they're used in the context of other calls
-AVALANCHEGO_VERSION=${AVALANCHEGO_VERSION:-$(getDepVersion github.com/ava-labs/avalanchego)}
+# TODO: undo this hack once go.mod is referring to a tag rather than a commit
+#AVALANCHEGO_VERSION=${AVALANCHEGO_VERSION:-$(getDepVersion github.com/ava-labs/avalanchego)}
+AVALANCHEGO_VERSION=${AVALANCHEGO_VERSION:-ab83fb41528de93c1790301cdd67a07dda9299f0}
```
[the remark this answers is not in the document]
Comment (feuGeneA) [reply scripts/versions.sh:22]: as discussed, we can't use the short commit hash that's present in go.mod. See CI failure at https://github.com/ava-labs/awm-relayer/actions/runs/10709822308/job/29695280372

File: signature-aggregator/aggregator/aggregator_test.go [hunk 10]
```diff
@@ -43,6 +44,7 @@ func instantiateAggregator(t *testing.T) (
 		1024,
 		sigAggMetrics,
 		messageCreator,
+		time.Now().Add(-1*time.Minute),
```
Comment (geoff-vball) [inline signature-aggregator/aggregator/aggregator_test.go:47]: What's the purpose of the -1 minute?

Comment (iansuvak) [reply signature-aggregator/aggregator/aggregator_test.go:47]: This is the `etnaTime` parameter. This would make the aggregator in this test run using the post-etna code paths since in this case `Etna` was activated a minute ago

File: scripts/versions.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (feuGeneA) [reply scripts/versions.sh:22]: For reference, to go with that CI link, here's the code change that it ran with: https://github.com/ava-labs/awm-relayer/commit/9a0224b67df1662d4a439b8bc10833f20a7b41e8

File: tests/etna_upgrade.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+	log.Info("Test Relaying from Subnet B to Subnet A")
+	testUtils.RelayBasicMessage(
+		ctx,
+		subnetBInfo,
+		subnetAInfo,
+		teleporterContractAddress,
+		fundedKey,
+		fundedAddress,
+	)
```
[the remark this answers is not in the document]
Comment (iansuvak) [reply tests/etna_upgrade.go:92]: Testing details done here are listed in the previous PR here: https://github.com/ava-labs/awm-relayer/pull/463

I guess a log message can be added and then logs captured from the relayer could be parsed but that would require changing how we start the binary which doesn't seem worth it for a relatively short lived test / piece of code.

[the remark this answers is not in the document]
Comment (geoff-vball) [reply tests/etna_upgrade.go:92]: Yeah, no need for big changes. Just wanted to make sure someone checked :)

File: relayer/config/config.go [hunk 12]
```diff
@@ -67,6 +68,9 @@ type Config struct {
 	DeciderURL             string                   `mapstructure:"decider-url" json:"decider-url"`
 	SignatureCacheSize     uint64                   `mapstructure:"signature-cache-size" json:"signature-cache-size"`
 
+	// mapstructure doesn't handle time.Time out of the box so handle it manually
+	EtnaTime time.Time `json:"etna-time"`
```
Comment (minghinmatthewlam) [inline relayer/config/config.go:72]: we could take in a string as well and use `time.Parse` in a date foramt

File: signature-aggregator/aggregator/aggregator.go [hunk 13]
```diff
@@ -590,3 +593,61 @@ func (s *SignatureAggregator) aggregateSignatures(
 	}
 	return aggSig, vdrBitSet, nil
 }
+
+// TODO: refactor this to remove special handling based on etnaTime
```
Comment (minghinmatthewlam) [inline signature-aggregator/aggregator/aggregator.go:597]: do we have an issue to track this

File: signature-aggregator/api/api.go [hunk 14]
```diff
@@ -94,25 +106,49 @@ func signatureAggregationAPIHandler(
...
+		message, err := types.UnpackWarpMessage(decodedMessage)
 		if err != nil {
 			msg := "Error unpacking warp message"
 			logger.Warn(msg, zap.Error(err))
 			writeJSONError(logger, w, msg)
 			return
 		}
+
+		justification, err := hex.DecodeString(
+			strings.TrimPrefix(req.Justification, "0x"),
```
Comment (minghinmatthewlam) [inline signature-aggregator/api/api.go:130]: we have `HexOrCB58ToID` and `SanitizieHexString` utils functions that might apply here

File: signature-aggregator/aggregator/aggregator.go (hunk 13 above)
Comment (iansuvak) [reply signature-aggregator/aggregator/aggregator.go:597]: Yes #466

File: tests/etna_upgrade.go [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+	// Set up relayer config
+	//
+	relayerConfig := testUtils.CreateDefaultRelayerConfig(
+		[]interfaces.SubnetTestInfo{subnetAInfo, subnetBInfo},
+		[]interfaces.SubnetTestInfo{subnetAInfo, subnetBInfo},
+		teleporterContractAddress,
+		fundedAddress,
+		relayerKey,
+	)
+	relayerConfig.EtnaTime = time.Now().AddDate(0, 0, -1)
```
Comment (cam-schultz) [inline tests/etna_upgrade.go:50]: Same here, a comment clarifying the -1 would be helpful

File: signature-aggregator/api/api.go (hunk 14 above)
Comment (iansuvak) [reply signature-aggregator/api/api.go:130]: This doesn't decode to an `ID` but yeah we should call `SanitizeHexString` here

File: relayer/config/config.go (hunk 12 above)
Comment (iansuvak) [reply relayer/config/config.go:72]: I like that the `v.GetTime()` call will handle both RFC3339 format as well as UNIX timestamps cleanly and output it unambiguously when being written out. 

With time.Parse we would need to manually handle conversion when writing it out to disk and reading it separately.

**Review by cam-schultz — APPROVED** [review]
A handful of final nits, but once they as well as the other open comments are addressed, this LGTM

File: tests/etna_upgrade.go (hunk 15 above)
Comment (iansuvak) [reply tests/etna_upgrade.go:51]: Addressed in 3d7707dc370ad953a957f6297b9f03cfd27c9431

File: signature-aggregator/api/api.go (hunk 14 above)
Comment (iansuvak) [reply signature-aggregator/api/api.go]: 3d7707dc370ad953a957f6297b9f03cfd27c9431

File: signature-aggregator/aggregator/aggregator_test.go (hunk 10 above)
Comment (iansuvak) [reply signature-aggregator/aggregator/aggregator_test.go:48]: 3d7707dc370ad953a957f6297b9f03cfd27c9431

**Review by cam-schultz — APPROVED** [review]

File: tests/utils/utils.go (hunk 3 above)
Comment (iansuvak) [reply tests/utils/utils.go]: Made the change back

**Review by geoff-vball — APPROVED** [review]

**Review by minghinmatthewlam — APPROVED** [review]
